File: LinkedList2D.py
from Node2D import Node2D


class LinkedList2D ():

    def __index__(self):
        self.first = None

    @classmethod
    def initialize(cls):
        ll = cls()
        return ll

    @classmethod
    def initializewithargs(cls, somenode):
        ll = cls()
        ll.first = somenode
        return ll

    def insertatbeginning(self, node):
        node.setNextNode(self.first)
        self.first = node

    def insertatend(self, argnode):
        if self.first == None:
            self.first = argnode
        else:
            tempnode = self.first
            while(tempnode.getnextnode() != None):
                tempnode = tempnode.getnextnode()
            tempnode.setnextnode(argnode)

    def removefirst(self):
        if self.first == None:
            return None
        else:
            tempnode = self.first
            self.first = self.first.getnextnode()
            return tempnode

    def removelast(self):
        if self.first == None:
            return None
        elif self.first.getnextnode() == None:
            tempnode = self.first
            self.first = None
            return tempnode
        else:
            tempnode1 = self.first
            tempnode2 = self.first.getnextnode()
            while tempnode2.getnextnode() != None:
                tempnode1 = tempnode2
                tempnode2 = tempnode2.getnextnode()
            tempnode1.setnextnode(None)
            return tempnode2

    def clear(self):
        self.first = None

File: Java/LinkedList2D.java
public class LinkedList2D {

    private Node2D first;

    /* default constructor: list is empty */
    LinkedList2D(){
        this.first = null;
    }

    /* overloaded constructor: list is set for an intial first node */
    LinkedList2D(Node2D first){
        this.first = first;
    }

    /* inserts a node at the beginning of the list */
    public void insertAtBeginning(Node2D node){
        node.setNextNode(first);
        first = node;
    }

    /* inserts a node at the end of the list */
    public void insertAtEnd(Node2D node) {
        if(first == null) {
            first = node;
        } else {
            Node2D temp = first;
            while(temp.getNextNode() != null) {
                temp = temp.getNextNode();
            }
            temp.setNextNode(node);
        }
    }

    /* deletes the first node in the list */
    public Node2D removeFirst(){
        if(first == null){
            return null;
        } else {
            Node2D temp = first;
            first = first.getNextNode();
            return temp;
        }
    }

    /* deletes the last node in the list */
    public Node2D removeLast(){
        if(first == null){
            return null;
        } else if (first.getNextNode() == null){
            Node2D temp = first;
            first = null;
            return temp;
        } else {
            Node2D temp1 = first;
            Node2D temp2 = first.getNextNode();
            while(temp2.getNextNode() != null){
                temp1 = temp2;
                temp2 = temp2.getNextNode();
            }
            temp1.setNextNode(null);
            return temp2;
        }
    }

    /* clears the list by setting first to null */
    public void clear(){
        first = null;
    }

}



